synapse/handlers/devicemessage.py

```diff
@@ -380,6 +387,12 @@ async def get_events_for_dehydrated_device(
             device_id,
         )
 
+        if messages == []:
+            # we've fetched all the messages, delete the dehydrated device
```

Comment:
This does not align with the MSC, the MSC tells us this:

> Once a client calls POST /dehydrated_device/{device_id}/events, the server can delete the device (though not necessarily its to-device messages).

So, per MSC, we should delete the device the first time the client makes a request to this endpoint, no matter how many messages we may have.

That being said, I think that we should leave the deletion of the device up to the client, see my rationale here: https://github.com/matrix-org/matrix-spec-proposals/pull/3814#discussion_r1266444656.